### PR TITLE
[NPUW] Fix memory consumption with bank uids

### PR DESCRIPTION
**Pull request by smirnov-alexey**
Introduced in https://github.com/openvinotoolkit/openvino/pull/28282

### CODE REVIEW

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 1]
```diff
@@ -70,6 +70,10 @@ ov::Tensor Bank::get(int64_t uid, const std::string& device) {
     // uid may be coming from a 2nd (3rd, ...) model
     // detach the tensor here just in case
     const_cast<LazyTensor&>(iter_device->second.lt).detach();
+    // Also detach from registered tensors
+    auto it_registered = device_bank.registered_tensors.find(iter_device->second.lt);
+    NPUW_ASSERT(it_registered != device_bank.registered_tensors.end());
+    const_cast<LazyTensor&>(it_registered->first).detach();
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:76]: Previously our `.get()` had effects. If the requested `LazyTensor` hasn't been evaluated yet, it is evaluated and copied to the device memory. And we did `.detach()` right after.

After the recent changes, `.get()` is just `.get()`. So here's the question - if there's no allocation happens at this point, why what has to be detached hasn't been detached at this point yet?

It seems we have too many `.detaches()` now - it just means our flow is not very clear.

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 2]
```diff
@@ -70,6 +70,10 @@ ov::Tensor Bank::get(int64_t uid, const std::string& device) {
     // uid may be coming from a 2nd (3rd, ...) model
     // detach the tensor here just in case
     const_cast<LazyTensor&>(iter_device->second.lt).detach();
+    // Also detach from registered tensors
+    auto it_registered = device_bank.registered_tensors.find(iter_device->second.lt);
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:74]: Why doesn't this happen on `evaluate_and_allocate()` ?

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp [hunk 3]
```diff
@@ -137,6 +141,11 @@ ov::Tensor Bank::eval_and_alloc(const LazyTensor& tensor,
     // Note: this is the non-CPU path!
     const_cast<LazyTensor&>(tensor).detach();
 
+    // Also detach from registered tensors
+    auto it_registered = dbank.registered_tensors.find(tensor);
+    NPUW_ASSERT(it_registered != dbank.registered_tensors.end());
+    const_cast<LazyTensor&>(it_registered->first).detach();
+
     return allocated_tensor;
```
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:149]: What we've had before the changes:

1. A `LazyTensor` was a key to the device tensor (for the transition from `lazy_closure` to `closure`)
2. Kvcache and Prefill models had its own individual LazyTensors (as distinct objects), populated from the original weights independently
3. When a compiled model queried a real tensor for the closure, it passed `LazyTensor` as a key and Bank called `.detatch()` in exchange
4. That is, the memory referred to twice was detached first in kvcache and then in prefill model, so finally cleared.

What do you have now:
1. A `LazyTensor` is a key to ID
2. First kvcache model exchanges its `LazyTensors` to IDs - this is where IDs are registered and LTs are collected inside
3. Then kvcache model evaluates the full bank - this is the point where LazyTensors should be detached
4. After that prefill model comes with its own `LazyTensors` and there's a hit, same IDs are returned
5. Evaluate on the bank essentially does nothing (and by the way you didn't optimize this part per past review).
6. We end up with memory leaked as the 2nd set of tensors was not detached.

How does detaching the existing registered tensors in the bank helps the problem? I believe this is what you catch in the second get, but in this case your internal storage still refers to the first set of LTs, not the second one.

The right solution is here:
1. Detach the LT in the bank once it is evaluated
2. On registerLT, detach the incoming LT if it was found it the bank. That's it.

**Review by dmatveev — CHANGES_REQUESTED** [review]

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 3 above)
Comment (dmatveev) [inline src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp:149]: This method has a return value but nobody is using it. I'd drop the body of this function completely - it will be easier to do all of this in the parent loop.

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 1 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 2 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Done

File: src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp (hunk 3 above)
Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Thanks for detailed comment. Done

Comment (smirnov-alexey) [reply src/plugins/intel_npu/src/plugin/npuw/weights_bank.cpp]: Done

**Review by dmatveev — APPROVED** [review]
